### PR TITLE
[WIP] Increase currently supported embedded/external databases granularity

### PR DESCRIPTION
**Pull request by miguelsorianod**
This PR performs a restructuration of the APIManager CRD to allow configurability at individual level for the databases that currently can be configured external so they can be individually selected as embedded or external instead of an all-or-nothing behavior, and also having configuration-level specific to each of them (in order to accommodate for new functionality that is specific to each DB).

Remarkable aspects of this PR:
* Deprecated the  `spec.system.database.mysql` and `spec.system.database.postgresql` sections. They have been replaced by the sections `spec.system.database.embedded` and `spec.system.database.external` (mutually exclusive) sections. Inside `spec.system.database.embeded` the `spec.system.database.embedded.mysql` and `spec.system.database.embedded.postgresql` sections have been added.
* Deprecated the fields `spec.system.redisImage`, `spec.backend.redisImage`
* Deprecated the highAvailability attribute that controlled whether all externally configurable DBs were external or not (in an all-or-nothing behavior for all of them) and it has been replaced by allowing each of those DBs to be configured as embedded or external individually
* Split the 'redis' component into SystemRedis and BackendRedis. The reason for this is that now we allow to control individually whether we want each of them being embedded or not.
* Added a CommonEmbeddedRedis component that includes the creation of data that is shared between SystemRedis and BackendRedis (at this moment, the 'redis-config' ConfigMap). This has resulted in additions/changes in the component, templates and operator packages.
* Upgrade procedure to migrate CR fields
* Renamed deprecated fields in CRD Go code with the 'Deprecated' prefix. This has changed the names at Go code level but NOT the json field names (to keep compatibility)
* This changes do NOT apply to templates. It is an operator-only functionality

### CODE REVIEW

**Comment by miguelsorianod** [comment]
@eguzki ready for a review

**Comment by miguelsorianod** [comment]
Some examples:

All DBs embbeded:

```
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: example-apimanager
spec:
  wildcardDomain: mywildcarddomain
  resourceRequirementsEnabled: true
  system:
    fileStorage:
      persistentVolumeClaim:
        storageClassName: nfs
```

All externally configurable DBs external:

```
kind: APIManager
metadata:
  name: example-apimanager
spec:
  wildcardDomain: mywildcarddomain
  resourceRequirementsEnabled: true
  system:
    fileStorage:
      persistentVolumeClaim:
        storageClassName: nfs
    database:
      external: {}
    redisDatabase:
      external: {}
  apicast:
  backend:
    redisDatabase:
      external: {}
```

Some externally configurable DBs external some embedded (system-redis external. system-database embedded postgresql and backend-redis embedded mysql)
```
kind: APIManager
metadata:
  name: example-apimanager
spec:
  wildcardDomain: mywildcarddomain
  resourceRequirementsEnabled: true
  system:
    fileStorage:
      persistentVolumeClaim:
        storageClassName: nfs
    redisDatabase:
      external: {}
    database:
      embedded:
        postgresql: {}
```

```
kind: APIManager
metadata:
  name: example-apimanager
spec:
  wildcardDomain: msoriano.apps.dev-eng-ocp4-3-backup-restore.dev.3sca.net
  resourceRequirementsEnabled: true
  system:
    fileStorage:
      persistentVolumeClaim:
        storageClassName: nfs
    image: registry.redhat.io/3scale-amp2/system-rhel7:3scale2.8
    database:
      embedded:
        postgresql:
          image: mysystempostgresqlimage
    redisDatabase:
      embedded:
        image: mysystemredisimage
  backend:
    redisDatabase:
      embedded:
        image: mybackendredisimage
```

File: deploy/crds/apps.3scale.net_apimanagers_crd.yaml [hunk 1]
```diff
@@ -107,13 +119,34 @@ spec:
...
+                            image:
+                              type: string
+                          type: object
+                        postgresql:
+                          properties:
+                            image:
+                              type: string
+                          type: object
+                      type: object
+                    external:
```
Comment (miguelsorianod) [inline deploy/crds/apps.3scale.net_apimanagers_crd.yaml:135]: in 'external' section I've added the mysql and postgresql section but I'm not sure whether we should differentiate in this case. In any case I've added them but I'm not using them in the code at the moment (in code I just check whether external is nil or not)

File: pkg/3scale/amp/auto-generated-templates/amp/amp-eval-s3.yml [hunk 2]
```diff
@@ -208,6 +208,62 @@ objects:
         type: ""
   status:
     dockerImageRepository: ""
+- apiVersion: v1
```
Comment (miguelsorianod) [inline pkg/3scale/amp/auto-generated-templates/amp/amp-eval-s3.yml:211]: Template changes that you'll observe should only be a change of order of the 'redis-config' CR object if I've implemented it correctly (due to the splitting of redis component in several components)

**Comment by eguzki** [comment]
this PR can be closed @miguelsorianod ?

**Comment by miguelsorianod** [comment]
We can close it as at least is not going to happen soon

**Pull request closed without merging**